internal/blueprint/fsnode_customizations.go

```diff
@@ -0,0 +1,468 @@
+	_, err := dir.ToFsNodeDirectory()
+	if err != nil {
+		return err
+	}
+
+	*d = dir
+	return nil
+}
+
+// Custom JSON unmarshalling for DirectoryCustomization with validation
+func (d *DirectoryCustomization) UnmarshalJSON(data []byte) error {
+	type directoryCustomization DirectoryCustomization
+
+	var dirPrivate directoryCustomization
+	if err := json.Unmarshal(data, &dirPrivate); err != nil {
+		return err
+	}
+
+	dir := DirectoryCustomization(dirPrivate)
+	if uid, ok := dir.User.(float64); ok {
+		// check if uid can be converted to int64
+		if uid != float64(int64(uid)) {
+			return fmt.Errorf("invalid user %f: must be an integer", uid)
+		}
+		dir.User = int64(uid)
+	}
+	if gid, ok := dir.Group.(float64); ok {
+		// check if gid can be converted to int64
+		if gid != float64(int64(gid)) {
+			return fmt.Errorf("invalid group %f: must be an integer", gid)
+		}
+		dir.Group = int64(gid)
+	}
+	// try converting to fsnode.Directory to validate all values
+	_, err := dir.ToFsNodeDirectory()
+	if err != nil {
+		return err
+	}
+
+	*d = dir
+	return nil
+}
+
+// ToFsNodeDirectory converts the DirectoryCustomization to an fsnode.Directory
+func (d DirectoryCustomization) ToFsNodeDirectory() (*fsnode.Directory, error) {
+	var mode *os.FileMode
+	if d.Mode != "" {
+		err := validateModeString(d.Mode)
+		if err != nil {
+			return nil, err
+		}
+		modeNum, err := strconv.ParseUint(d.Mode, 8, 32)
+		if err != nil {
+			return nil, fmt.Errorf("invalid mode %s: %v", d.Mode, err)
+		}
+		mode = common.ToPtr(os.FileMode(modeNum))
+	}
+
+	return fsnode.NewDirectory(d.Path, mode, d.User, d.Group, d.EnsureParents)
+}
+
+// DirectoryCustomizationsToFsNodeDirectories converts a slice of DirectoryCustomizations
+// to a slice of fsnode.Directories
+func DirectoryCustomizationsToFsNodeDirectories(dirs []DirectoryCustomization) ([]*fsnode.Directory, error) {
+	if len(dirs) == 0 {
+		return nil, nil
+	}
+
+	var fsDirs []*fsnode.Directory
+	var errors []error
+	for _, dir := range dirs {
+		fsDir, err := dir.ToFsNodeDirectory()
+		if err != nil {
+			errors = append(errors, err)
+		}
+		fsDirs = append(fsDirs, fsDir)
+	}
+
+	if len(errors) > 0 {
+		return nil, fmt.Errorf("invalid directory customizations: %v", errors)
+	}
+
+	return fsDirs, nil
+}
+
+// FileCustomization represents a file to be created in the image
+type FileCustomization struct {
+	// Absolute path to the file
+	Path string `json:"path" toml:"path"`
+	// Owner of the directory specified as a string (user name), int64 (UID) or nil
+	User interface{} `json:"user,omitempty" toml:"user,omitempty"`
+	// Owner of the directory specified as a string (group name), int64 (UID) or nil
+	Group interface{} `json:"group,omitempty" toml:"group,omitempty"`
+	// Permissions of the file specified as an octal number
+	Mode string `json:"mode,omitempty" toml:"mode,omitempty"`
+	// Data is the file content in plain text
+	Data string `json:"data,omitempty" toml:"data,omitempty"`
```

Comment:
I would prefer to be able to have both `DataAscii` and `DataBytes` to be able to add binary files, we can represent the latter as one of the base encodings in the json/toml?

Reply:
I'm with you on this, but the requirement is to support only plain text data for MVP.

As I mentioned in the description, adding support for base64 encoded (potentially binary) data would be super simple, since internaly it's all bytes. My plan would be to later add `data_base64` property later on to hold these. Since the expectation is that users will more often use the "plain text" input, I didn't add "Ascii" into the property name. There is nothing blocking us from renaming the Structure member name to `DataAscii` once the support for base64 encoded content is added, but I didn't see value in doing so now.

Reply:
Might also make sense to keep just `Data` and have a toggle for `Binary`?  Or an enum encoding/format, but probably not as simple.  A boolean for `Binary = true/false` would make it easier to handle a single data field without needing to check or tell users that they can't use both.